File: matrix.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

const n = 2048

func main() {
	a := make([][]float32, n)
	for i := range a {
		a[i] = make([]float32, n)
	}
	b := make([][]float32, n)
	for i := range b {
		b[i] = make([]float32, n)
	}
	c := make([][]float32, n)
	for i := range c {
		c[i] = make([]float32, n)
	}

	for i := 0; i < n; i++ {
		rand.Seed(time.Now().UnixNano())

		for j := 0; j < n; j++ {
			a[i][j] = rand.Float32()
			b[i][j] = rand.Float32()
		}
	}

	st := time.Now()

	for i := 0; i < n; i++ {
		for j := 0; j < n; j++ {
			for k := 0; k < n; k++ {
				c[i][j] += a[i][k] * b[j][k]
			}
		}
	}

	fmt.Println(time.Since(st).Seconds())
}

File: matrix.rs
use std::time::{SystemTime, Instant, UNIX_EPOCH};

const N: usize = 2048;

// minimal crate-free Xorshift 128 random number generator

const KX: u32 = 123456789;
const KY: u32 = 362436069;
const KZ: u32 = 521288629;
const KW: u32 = 88675123;

pub struct Rand {
    x: u32,
    y: u32,
    z: u32,
    w: u32,
}

impl Rand {
    pub fn new(seed: u32) -> Rand {
        Rand {
            x: KX ^ seed,
            y: KY ^ seed,
            z: KZ,
            w: KW,
        }
    }

    pub fn rand(&mut self) -> u32 {
        let t = self.x ^ self.x.wrapping_shl(11);
        self.x = self.y;
        self.y = self.z;
        self.z = self.w;
        self.w ^= self.w.wrapping_shr(19) ^ t ^ t.wrapping_shr(8);
        return self.w;
    }

    pub fn rand_float(&mut self) -> f64 {
        (self.rand() as f64) / (<u32>::max_value() as f64)
    }
}

fn main() {
    let mut rng = Rand::new(SystemTime::now().duration_since(UNIX_EPOCH).unwrap().as_secs_f32() as u32);
    let mut a: Vec<Vec<f64>> = vec![vec![0.0; N]; N];
    let mut b: Vec<Vec<f64>> = vec![vec![0.0; N]; N];
    let mut c: Vec<Vec<f64>> = vec![vec![0.0; N]; N];

    for i in 0..N {
        for j in 0..N {
            a[i][j] = rng.rand_float();
            b[i][j] = rng.rand_float();
        }
    }

    let nw = Instant::now();

    for i in 0..N {
        for j in 0..N {
            for k in 0..N {
                c[i][j] += a[i][k] * b[k][j];
            }
        }
    }

    println!("{}", nw.elapsed().as_secs());
}
